### PR TITLE
api endpoints for request messages

### PR DESCRIPTION
**Pull request by ghost**
add python-dotenv to requirements.txt
api endpoint for sending message requests
api endpoint for accepting message requests
api endpoint for rejecting message requests
api endpoint for get all message requests

### CODE REVIEW

File: message_app/db/db.py [hunk 1]
```diff
@@ -11,11 +11,14 @@
...
     if 'db' not in g:
+        print(current_app.config['DATABASE'])
         g.db = sqlite3.connect(
             current_app.config['DATABASE'],
             detect_types=sqlite3.PARSE_DECLTYPES
         )
-        g.db.row_factory = sqlite3.Row
+        # Enable validating foreign_keys
+        g.db.execute("PRAGMA foreign_keys = ON;")
+        g.db.row_factory = make_dicts
```
Comment (khuolo01) [inline message_app/db/db.py:49]: Is this line any different than the line you deleted above (**_g.db.row_factory = sqlite3.Row_**)? I think _**sqlite3.Row**_ is doing the same thing here.

File: message_app/db/db.py [hunk 2]
```diff
@@ -40,6 +43,10 @@ def query_db(query: str, args={}, one=False):
     cur.close()
     return (rv[0] if rv else None) if one else rv
 
+def make_dicts(cursor, row):
```
Comment (khuolo01) [inline message_app/db/db.py:74]: Is this function the same with _**sqlite3.Row**_ ?

File: tests/conftest.py [hunk 3]
```diff
@@ -18,7 +22,7 @@ def app():
     })
 
     with app.app_context():
-        init_db()
+        get_db().executescript(_data_sql)
```
Comment (khuolo01) [inline tests/conftest.py]: I see you are populating the database with test data, which means that you still have to connect to the database later on. Should we make this connection to the database in unit testing? I feel like we should mock the values directly with the populated data instead of making real connection.

File: message_app/__init__.py [hunk 4]
```diff
@@ -8,7 +8,7 @@ def create_app(test_config=None):
     app = Flask(__name__)
     app.config.from_mapping(
         SECRET_KEY='dev',
-        DATABASE=os.path.join(app.instance_path, "db", 'message_app_db.sqlite3'),
+        DATABASE=os.path.join("message_app/db", 'message_app_db.sqlite3'),
```
Comment (leth02) [inline message_app/__init__.py]: Why is this changed to hard coded path `"message_app/db"`?

File: message_app/db/db.py [hunk 5]
```diff
@@ -40,6 +42,10 @@ def query_db(query: str, args={}, one=False):
...
+    """convert the retrieved data into dictionary with key is column name"""
```
Comment (leth02) [inline message_app/db/db.py:75]: `make_dicts` usage will be fully replaced with SQLAlchemy in the future right?

File: message_app/request_message.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
+from flask import Blueprint, render_template, jsonify
+from message_app.db import db
+
+request_messages = Blueprint("request_messages", __name__)
+
+@request_messages.route("/api/request/send/<int:sender_id>/<int:receiver_id>/<int:request_time>", methods=["POST"])
+def send_request(sender_id, receiver_id, request_time):
+    try:
+        requestData = db.query_db(
```
Comment (leth02) [inline message_app/request_message.py]: please use snake case for all occurrences of `requestData`

File: tests/test_request_message.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+@pytest.mark.parametrize(("user_id", "status_code", "response_message"), (
+    (1004, 400, b'{"Error":"Bad Request.No user found"}\n'),
+    (1005, 400, b'{"Error":"Bad Request.No user found"}\n'),
+    (1006, 400, b'{"Error":"Bad Request.No user found"}\n')
+    ))
+def test_get_all_requests_fail(client, app, user_id, status_code, response_message):
+    response = client.get(f"/api/request/all/{user_id}")
+    assert response.status_code == status_code
+    assert response.data == response_message
+
```
Comment (leth02) [inline tests/test_request_message.py]: You should group them into a Python class.

Also, with each assert statement, please add a more explanatory/human readable description of why the test fails (where relevant). For example:
`assert response.data == "some data", "Fetching all conversation requests failed"`

**Review by leth02 — COMMENTED** [review]
I have suggested some minor changes.

File: message_app/__init__.py (hunk 4 above)
Comment (ghost) [reply message_app/__init__.py]: `app.instance_path` is a path to instance folder in the project directory, and we do not have that folder. In the tutorial, they create database in instance folder.
https://uniwebsidad.com/libros/explore-flask/chapter-5/instance-folder

File: message_app/db/db.py (hunk 5 above)
Comment (ghost) [reply message_app/db/db.py:75]: yes, after we have ORM, it can be remove

**Review by khuolo01 — CHANGES_REQUESTED** [review]

File: tests/test_request_message.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    (1000, 1002, 123456, 200, b'Success'),
+    ))
+def test_sending_request(client, app, sender_id, receiver_id, request_time, status_code, response_message):
+
+    response = client.post(f"/api/request/send/{sender_id}/{receiver_id}/{request_time}")
+    assert response.status_code == status_code
+    assert response.data == response_message
+
+    with app.app_context():
+        assert query_db(
```
Comment (khuolo01) [inline tests/test_request_message.py:16]: You should mock the connection.

File: message_app/db/db.py (hunk 1 above)
Comment (ghost) [reply message_app/db/db.py:49]: sqlite3.Row return tuple, not dict

**Review by khuolo01 — CHANGES_REQUESTED** [review]
You should mock all the connections to the database.

File: tests/conftest.py (hunk 3 above)
Comment (leth02) [reply tests/conftest.py]: As we have discussed that we will populate a test database for testing APIs that interact with the database (the purpose of the endpoint is mainly to interact with the database), I will need to add a PR to use a real test database before we can progress here, right? @khuolo01

Comment (khuolo01) [reply tests/conftest.py]: My new PR has tests involving creating a test database before the test, populating the test database during testing, and deleting the test database after testing, so I don't think we need a separate PR to use a real test database.

**Comment by leth02** [comment]
@vuongtranluc Could you rebase and resolve the conflicts before I can give the final review?

**Review by leth02 — APPROVED** [review]
Just a reminder: There will be another PR where Luc will change the API tests from using raw sql commands to using models' methods.

**Review by khuolo01 — APPROVED** [review]
LGTM
